src/macros.rs

```diff
@@ -720,3 +720,53 @@ macro_rules! const_unreachable {
         loop {}
     }};
 }
+
+/// Asserts at compile time that `$condition` is true for `Self` or the given
+/// `$tyvar`s.
+macro_rules! ct_assert {
+    (Self $(: $(? $optbound:ident $(+)?)* $($bound:ident $(+)?)* )? => $condition:expr) => {
+        trait ConstAssert {
+            const ASSERT: bool;
+        }
+
+        impl<T $(: $(? $optbound +)* $($bound +)*)?> ConstAssert for T {
+            const ASSERT: bool = {
+                const_assert!($condition);
+                $condition
+            };
+        }
+
+        const_assert!(<Self as ConstAssert>::ASSERT);
+    };
+    ($($tyvar:ident $(: $(? $optbound:ident $(+)?)* $($bound:ident $(+)?)* )?),* => $condition:expr) => {
+        trait ConstAssert {
+            const ASSERT: bool;
+        }
+
+        impl<$($tyvar $(: $(? $optbound +)* $($bound +)*)?,)*> ConstAssert for ($($tyvar,)*) {
+            const ASSERT: bool = {
+                const_assert!($condition);
+                $condition
+            };
+        }
+
+        const_assert!(<($($tyvar,)*) as ConstAssert>::ASSERT);
+    };
+}
+
+/// Assert at compile time that `tyvar` does not have a zero-sized DST
+/// component.
+macro_rules! ct_assert_dst_is_not_zst {
+    ($tyvar:ident) => {
+        use crate::KnownLayout;
+        ct_assert!($tyvar: ?Sized + KnownLayout => {
+            let dst_is_zst = match $tyvar::LAYOUT.size_info {
+                crate::SizeInfo::Sized { .. } => false,
+                crate::SizeInfo::SliceDst(TrailingSliceLayout { elem_size, .. }) => {
+                    elem_size == 0
+                }
+            };
+            !dst_is_zst
+        });
+    }
```

Comment:
Wrap this in an extra set of curly braces? Otherwise, `use crate::KnownLayout;` is introduced into the caller's namespace.

Reply:
[Done](https://github.com/google/zerocopy/compare/e36690aa8bf91b56899b9ad2b70eeb5f75a49362..2a433d49b23b4eba40cf8843490b0940b24894ac#diff-315c02cd05738da173861537577d159833f70f79cfda8cd7cf1a0d7a28ace31bR760)